### ISSUE TITLE
xref.Decorations: add mode to retrieve all references that contain a span [phab:D811]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1106]
_Created by **schroederc** at 2016-04-04 21:05:03_: 

This addresses use-cases similiar to `kwazthis` where a client would like
to know what decorations are **at** a particular point or span rather
than **within** a span.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-04-04 21:36:58_:

"any decorations that…are" or "any decoration that…is" (your choice :)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-04-04 21:36:58_:

"At" seems misleading. `AROUND_SPAN` perhaps?  (Other possibilities: "enclosing", "surrounding", "containing", <fill in the blank>)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-04-04 21:36:58_:

I suggest a helper for this:

```
func anchorInBounds(kind xpb.DecorationsRequest_SpanKind, start, end, startBounds, endBounds int) bool {
	switch kind {
	case xpb.DecorationsRequest_WITHIN_SPAN:
		return start >= startBounds && end <= endBounds
	case xpb.DecorationsRequest_AROUND_SPAN:
		return start <= startBounds && end >= endBounds
	}
	return false
}
```

Not that it's a lot of code, but it's a fiddly condition. (I'd probably log an error in the fallback case, too, but YMMV).

**Comment by kythe** [comment]
_**schroederc** wrote at 2016-04-05 21:42:46_:

Review changes

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-04-05 21:45:35_:

Lets do `AROUND_SPAN`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-04-05 21:48:25_:

surround

**Comment by kythe** [comment]
_**fromberger** accepted this revision at 2016-04-05 21:48:25_

**Comment by kythe** [comment]
_**schroederc** wrote at 2016-04-05 21:51:01_:

Typo

**Comment by kythe** [comment]
_**schroederc** closed via commit c087e029ca19e0a597fe11d5051349c68a00b30a at 2016-04-05 21:52:11_
